easybuild/easyblocks/o/openssl_wrapper.py

```diff
@@ -0,0 +1,269 @@
+from easybuild.tools.run import run_cmd
+from easybuild.tools.systemtools import DARWIN, LINUX, get_os_type, get_shared_lib_ext, find_library_path
+from easybuild.tools.build_log import EasyBuildError, print_warning
+
+
+class EB_OpenSSL_wrapper(Bundle):
+    """
+    Locate the installation files of OpenSSL in the host system
+    If available, wrap the system OpenSSL by symlinking all installation files
+    Fall back to the bundled component otherwise
+    """
+
+    @staticmethod
+    def extra_options(extra_vars=None):
+        """Easyconfig parameters specific to OpenSSL wrapper"""
+        extra_vars = Bundle.extra_options(extra_vars=extra_vars)
+        extra_vars.update({
+            'wrap_system_openssl': [True, 'Detect and wrap OpenSSL installation in host system', CUSTOM],
+        })
+        return extra_vars
+
+    def __init__(self, *args, **kwargs):
+        """Locate the installation files of OpenSSL in the host system"""
+        super(EB_OpenSSL_wrapper, self).__init__(*args, **kwargs)
+
+        # Libraries packaged in OpenSSL
+        openssl_libs = ['libssl', 'libcrypto']
+        # list of relevant library extensions per system and version of OpenSSL
+        # the first item should be the extension of an installation from source,
+        # it will be used in the sanity checks of the component
+        openssl_libext = {
+            '1.0': {
+                LINUX: ('so.1.0.0', 'so.10'),
+                DARWIN: ('1.0.dylib', ),
+            },
+            '1.1': {
+                LINUX: ('so.1.1', ),
+                DARWIN: ('1.1.dylib', ),
+            },
+        }
+
+        os_type = get_os_type()
+        if self.version in openssl_libext and os_type in openssl_libext[self.version]:
+            # generate matrix of versioned .so filenames
+            self.openssl_libs = [
+                ['%s.%s' % (lib, ext) for ext in openssl_libext[self.version][os_type]]
+                for lib in openssl_libs
+            ]
+        else:
+            raise EasyBuildError("Don't know name of OpenSSL system library for version %s and OS type %s",
+                                 self.version, os_type)
+
+        # folders containing engines libraries
+        self.openssl_engines = {
+            '1.0': 'engines',
+            '1.1': 'engines-1.1',
+        }
+
+        # Paths to system libraries and headers of OpenSSL
+        self.ssl_syslib = None
+        self.ssl_sysheader = None
+        self.ssl_sysengines = None
+        self.ssl_sysbin = None
+
+        if not self.cfg.get('wrap_system_openssl'):
+            return
+
+        # Check the system libraries of OpenSSL
+        for n, libssl in enumerate(self.openssl_libs[0]):
+            self.ssl_syslib = find_library_path(libssl)
+            if self.ssl_syslib:
+                # reduce matrix of library names to this one
+                self.openssl_libs = zip(*self.openssl_libs)[n]
```

Comment:
This looks weird, and it doesn't work either (at least not in Python 3):
```
ERROR: Traceback (most recent call last):
  File "/users/kehoste/easybuild-framework/easybuild/main.py", line 117, in build_and_install_software
    (ec_res['success'], app_log, err) = build_and_install_one(ec, init_env)
  File "/users/kehoste/easybuild-framework/easybuild/framework/easyblock.py", line 3533, in build_and_install_one
    app = app_class(ecdict['ec'])
  File "/tmp/eb-a0bcx10h/included-easyblocks-y6ba563p/easybuild/easyblocks/openssl_wrapper.py", line 108, in __init__
    self.openssl_libs = zip(*self.openssl_libs)[n]
TypeError: 'zip' object is not subscriptable
```

Reply:
right, I forgot about python3, `zip` returns an iterator instead of a list. Fixed in https://github.com/easybuilders/easybuild-easyblocks/pull/2429/commits/724ea6ec2fe2bc2b99e0c74a56820056056edf90

Apart from that, I don't know how to make it less weird, it's just an unpack into `zip`. Maybe I could take the element from the list in another line, something like
```
zipped_libs = list(zip(*self.openssl_libs))
self.openssl_libs = zipped_libs[idx]
```

Reply:
actually, we can just use a list comprehension. Done in https://github.com/easybuilders/easybuild-easyblocks/pull/2429/commits/9a9f34a64f77d05e2e6a3e6c5d2d8688c164ad25